In [1]:
#Importing all the necessary libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import datetime as dt

%matplotlib inline

In [2]:
#Load data
df = pd.read_csv("processed_data.csv")
df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                         review_id  \
0           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
1           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
2           2017-10-18 00:00:00  a54f0611adc9ed256b57ede6b6eb5114   
3           2018-08-13 00:00:00  8d5266042046a06655c8db133d120ba5   
4           2018-09-04 00:00:00  e73b67b67587f7644d5bd1a52deb1b01   

   review_score  ... product_description_lenght product_photos_qty  \
0             4  ...                      268.0                4.0   
1             4  ...                      268.0                4.0   
2             4  ...                      268.0                4.0   
3             4  ...                      178.0                1.0   
4             5  ...                      232.0                1.0   

  product_weight_g product_length_cm  product_height_cm product_width_cm  \
0            500.0              19.0                8.0             13.0   
1            500.0              19.0                8.0             13.0   
2            500.0              19.0                8.0             13.0   
3            400.0              19.0               13.0             19.0   
4            420.0              24.0               19.0             21.0   

                 customer_unique_id  customer_zip_code_prefix  customer_city  \
0  7c396fd4830fd04220f754e42b4e5bff                      3149      sao paulo   
1  7c396fd4830fd04220f754e42b4e5bff                      3149      sao paulo   
2  7c396fd4830fd04220f754e42b4e5bff                      3149      sao paulo   
3  af07308b275d755c9edb36a90c618231                     47813      barreiras   
4  3a653a41f6f9fc3d2a113cf8398680e8                     75265     vianopolis   

  customer_state  
0             SP  
1             SP  
2             SP  
3             BA  
4             GO  

[5 rows x 39 columns]

In [3]:
df.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'order_item_id', 'product_id',
       'seller_id', 'shipping_limit_date', 'price', 'freight_value',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'customer_unique_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state'],
      dtype='object')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119151 entries, 0 to 119150
Data columns (total 39 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       119151 non-null  object 
 1   customer_id                    119151 non-null  object 
 2   order_status                   119151 non-null  object 
 3   order_purchase_timestamp       119151 non-null  object 
 4   order_approved_at              118974 non-null  object 
 5   order_delivered_carrier_date   117065 non-null  object 
 6   order_delivered_customer_date  115730 non-null  object 
 7   order_estimated_delivery_date  119151 non-null  object 
 8   review_id                      119151 non-null  object 
 9   review_score                   119151 non-null  int64  
 10  review_comment_title           14189 non-null   object 
 11  review_comment_message         51250 non-null   object 
 12  review_creation_date          

In [5]:
# Price variable
df.price

0          29.99
1          29.99
2          29.99
3         118.70
4         159.90
           ...  
119146    174.90
119147    205.99
119148    179.99
119149    179.99
119150     68.50
Name: price, Length: 119151, dtype: float64

In [6]:
df.customer_id.nunique()

99441

In [7]:
df.customer_id.count()

119151

In [8]:
df.product_id.nunique()

32951

In [9]:
df.seller_id.nunique()                      

3095

In [10]:
df.order_item_id.count()                      

118318

In [11]:
x = list(set(df.order_id))
dic = dict(zip(x, list(range(1,len(x)+1))))


In [12]:
df.order_id

0         e481f51cbdc54678b7cc49136f2d6af7
1         e481f51cbdc54678b7cc49136f2d6af7
2         e481f51cbdc54678b7cc49136f2d6af7
3         53cdb2fc8bc7dce0b6741e2150273451
4         47770eb9100c2d0c44946d9cf07ec65d
                        ...               
119146    63943bddc261676b46f01ca7ac2f7bd8
119147    83c1379a015df1e13d02aae0204711ab
119148    11c177c8e97725db2631073c19f07b62
119149    11c177c8e97725db2631073c19f07b62
119150    66dea50a8b16d9b4dee7af250b4be1a5
Name: order_id, Length: 119151, dtype: object

In [13]:
len(dic)

99441

In [14]:
len(x)

99441

In [15]:
df[['payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'order_item_id', 'product_id',
       'seller_id', 'shipping_limit_date', 'price', 'freight_value']]

payment_sequential payment_type  payment_installments  payment_value  \
0                      1.0  credit_card                   1.0          18.12   
1                      3.0      voucher                   1.0           2.00   
2                      2.0      voucher                   1.0          18.59   
3                      1.0       boleto                   1.0         141.46   
4                      1.0  credit_card                   3.0         179.12   
...                    ...          ...                   ...            ...   
119146                 1.0  credit_card                   3.0         195.00   
119147                 1.0  credit_card                   5.0         271.01   
119148                 1.0  credit_card                   4.0         441.16   
119149                 1.0  credit_card                   4.0         441.16   
119150                 1.0   debit_card                   1.0          86.86   

        order_item_id                        product_id  \
0                 1.0  87285b34884572647811a353c7ac498a   
1                 1.0  87285b34884572647811a353c7ac498a   
2                 1.0  87285b34884572647811a353c7ac498a   
3                 1.0  595fac2a385ac33a80bd5114aec74eb8   
4                 1.0  aa4383b373c6aca5d8797843e5594415   
...               ...                               ...   
119146            1.0  f1d4ce8c6dd66c47bbaa8c6781c2a923   
119147            1.0  b80910977a37536adeddd63663f916ad   
119148            1.0  d1c427060a0f73f6b889a5c7c61f2ac4   
119149            2.0  d1c427060a0f73f6b889a5c7c61f2ac4   
119150            1.0  006619bbed68b000c8ba3f8725d5409e   

                               seller_id  shipping_limit_date   price  \
0       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
1       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
2       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
3       289cdb325fb7e7f891c38608bf9e0962  2018-07-30 03:24:27  118.70   
4       4869f7a5dfa277a7dca6462dcf3b52b2  2018-08-13 08:55:23  159.90   
...                                  ...                  ...     ...   
119146  1f9ab4708f3056ede07124aad39a2554  2018-02-12 13:10:37  174.90   
119147  d50d79cb34e38265a8649c383dcffd48  2017-09-05 15:04:16  205.99   
119148  a1043bafd471dff536d0c462352beb48  2018-01-12 21:36:21  179.99   
119149  a1043bafd471dff536d0c462352beb48  2018-01-12 21:36:21  179.99   
119150  ececbfcff9804a2d6b40f589df8eef2b  2018-03-15 10:55:42   68.50   

        freight_value  
0                8.72  
1                8.72  
2                8.72  
3               22.76  
4               19.22  
...               ...  
119146          20.10  
119147          65.02  
119148          40.59  
119149          40.59  
119150          18.36  

[119151 rows x 10 columns]

In [16]:
df[['seller_id', 'shipping_limit_date', 'price', 'freight_value',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght']]

seller_id  shipping_limit_date   price  \
0       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
1       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
2       3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99   
3       289cdb325fb7e7f891c38608bf9e0962  2018-07-30 03:24:27  118.70   
4       4869f7a5dfa277a7dca6462dcf3b52b2  2018-08-13 08:55:23  159.90   
...                                  ...                  ...     ...   
119146  1f9ab4708f3056ede07124aad39a2554  2018-02-12 13:10:37  174.90   
119147  d50d79cb34e38265a8649c383dcffd48  2017-09-05 15:04:16  205.99   
119148  a1043bafd471dff536d0c462352beb48  2018-01-12 21:36:21  179.99   
119149  a1043bafd471dff536d0c462352beb48  2018-01-12 21:36:21  179.99   
119150  ececbfcff9804a2d6b40f589df8eef2b  2018-03-15 10:55:42   68.50   

        freight_value  seller_zip_code_prefix     seller_city seller_state  \
0                8.72                  9350.0            maua           SP   
1                8.72                  9350.0            maua           SP   
2                8.72                  9350.0            maua           SP   
3               22.76                 31570.0  belo horizonte           SP   
4               19.22                 14840.0         guariba           SP   
...               ...                     ...             ...          ...   
119146          20.10                 17602.0            tupa           SP   
119147          65.02                  8290.0       sao paulo           SP   
119148          40.59                 37175.0        ilicinea           MG   
119149          40.59                 37175.0        ilicinea           MG   
119150          18.36                 14407.0          franca           SP   

         product_category_name  product_name_lenght  
0        utilidades_domesticas                 40.0  
1        utilidades_domesticas                 40.0  
2        utilidades_domesticas                 40.0  
3                   perfumaria                 29.0  
4                   automotivo                 46.0  
...                        ...                  ...  
119146                   bebes                 52.0  
119147      eletrodomesticos_2                 51.0  
119148  informatica_acessorios                 59.0  
119149  informatica_acessorios                 59.0  
119150            beleza_saude                 45.0  

[119151 rows x 9 columns]

In [17]:
df[['product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'customer_unique_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state']]

product_description_lenght  product_photos_qty  product_weight_g  \
0                            268.0                 4.0             500.0   
1                            268.0                 4.0             500.0   
2                            268.0                 4.0             500.0   
3                            178.0                 1.0             400.0   
4                            232.0                 1.0             420.0   
...                            ...                 ...               ...   
119146                       828.0                 4.0            4950.0   
119147                       500.0                 2.0           13300.0   
119148                      1893.0                 1.0            6550.0   
119149                      1893.0                 1.0            6550.0   
119150                       569.0                 1.0             150.0   

        product_length_cm  product_height_cm  product_width_cm  \
0                    19.0                8.0              13.0   
1                    19.0                8.0              13.0   
2                    19.0                8.0              13.0   
3                    19.0               13.0              19.0   
4                    24.0               19.0              21.0   
...                   ...                ...               ...   
119146               40.0               10.0              40.0   
119147               32.0               90.0              22.0   
119148               20.0               20.0              20.0   
119149               20.0               20.0              20.0   
119150               16.0                7.0              15.0   

                      customer_unique_id  customer_zip_code_prefix  \
0       7c396fd4830fd04220f754e42b4e5bff                      3149   
1       7c396fd4830fd04220f754e42b4e5bff                      3149   
2       7c396fd4830fd04220f754e42b4e5bff                      3149   
3       af07308b275d755c9edb36a90c618231                     47813   
4       3a653a41f6f9fc3d2a113cf8398680e8                     75265   
...                                  ...                       ...   
119146  da62f9e57a76d978d02ab5362c509660                     11722   
119147  737520a9aad80b3fbbdad19b66b37b30                     45920   
119148  5097a5312c8b157bb7be58ae360ef43c                     28685   
119149  5097a5312c8b157bb7be58ae360ef43c                     28685   
119150  60350aa974b26ff12caad89e55993bd6                     83750   

       customer_city customer_state  
0          sao paulo             SP  
1          sao paulo             SP  
2          sao paulo             SP  
3          barreiras             BA  
4         vianopolis             GO  
...              ...            ...  
119146  praia grande             SP  
119147   nova vicosa             BA  
119148       japuiba             RJ  
119149       japuiba             RJ  
119150          lapa             PR  

[119151 rows x 10 columns]

In [18]:
df.drop(['customer_id', 'order_id','review_id', 'review_comment_title','review_comment_message', 'review_creation_date',
'seller_city', 'shipping_limit_date', 'customer_unique_id', 'customer_city'], axis=1,inplace=True)

In [19]:
df.head()

order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  review_score review_answer_timestamp  \
0           2017-10-18 00:00:00             4     2017-10-12 03:43:48   
1           2017-10-18 00:00:00             4     2017-10-12 03:43:48   
2           2017-10-18 00:00:00             4     2017-10-12 03:43:48   
3           2018-08-13 00:00:00             4     2018-08-08 18:37:50   
4           2018-09-04 00:00:00             5     2018-08-22 19:07:58   

   payment_sequential payment_type  ...  product_category_name  \
0                 1.0  credit_card  ...  utilidades_domesticas   
1                 3.0      voucher  ...  utilidades_domesticas   
2                 2.0      voucher  ...  utilidades_domesticas   
3                 1.0       boleto  ...             perfumaria   
4                 1.0  credit_card  ...             automotivo   

   product_name_lenght  product_description_lenght product_photos_qty  \
0                 40.0                       268.0                4.0   
1                 40.0                       268.0                4.0   
2                 40.0                       268.0                4.0   
3                 29.0                       178.0                1.0   
4                 46.0                       232.0                1.0   

  product_weight_g  product_length_cm  product_height_cm  product_width_cm  \
0            500.0               19.0                8.0              13.0   
1            500.0               19.0                8.0              13.0   
2            500.0               19.0                8.0              13.0   
3            400.0               19.0               13.0              19.0   
4            420.0               24.0               19.0              21.0   

  customer_zip_code_prefix customer_state  
0                     3149             SP  
1                     3149             SP  
2                     3149             SP  
3                    47813             BA  
4                    75265             GO  

[5 rows x 29 columns]

In [20]:
df.columns

Index(['order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'review_score',
       'review_answer_timestamp', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'order_item_id', 'product_id',
       'seller_id', 'price', 'freight_value', 'seller_zip_code_prefix',
       'seller_state', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'customer_zip_code_prefix', 'customer_state'],
      dtype='object')

In [21]:
df.order_status.unique()

array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

In [22]:
df.order_status.value_counts()#.plot(kind="bar")

delivered      115731
shipped          1256
canceled          750
unavailable       652
invoiced          378
processing        376
created             5
approved            3
Name: order_status, dtype: int64

In [23]:
df.order_status.value_counts().index

Index(['delivered', 'shipped', 'canceled', 'unavailable', 'invoiced',
       'processing', 'created', 'approved'],
      dtype='object')

In [24]:
# Creating a map dict for oredr status
order_status_dict = {'delivered':8, 'shipped':7, 'canceled':6, 'unavailable':5, 'invoiced':4,
       'processing':3, 'created':2, 'approved':1}

In [25]:
# mapping the status variable
df.order_status = df.order_status.map(order_status_dict)
df.head(2)

order_status order_purchase_timestamp    order_approved_at  \
0             8      2017-10-02 10:56:33  2017-10-02 11:07:15   
1             8      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date  review_score review_answer_timestamp  \
0           2017-10-18 00:00:00             4     2017-10-12 03:43:48   
1           2017-10-18 00:00:00             4     2017-10-12 03:43:48   

   payment_sequential payment_type  ...  product_category_name  \
0                 1.0  credit_card  ...  utilidades_domesticas   
1                 3.0      voucher  ...  utilidades_domesticas   

   product_name_lenght  product_description_lenght product_photos_qty  \
0                 40.0                       268.0                4.0   
1                 40.0                       268.0                4.0   

  product_weight_g  product_length_cm  product_height_cm  product_width_cm  \
0            500.0               19.0                8.0              13.0   
1            500.0               19.0                8.0              13.0   

  customer_zip_code_prefix customer_state  
0                     3149             SP  
1                     3149             SP  

[2 rows x 29 columns]

In [26]:
df[['order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date']] = df[['order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date']].apply(pd.to_datetime)

In [27]:
df.order_approved_at.dtype

dtype('<M8[ns]')

In [28]:
df.columns

Index(['order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'review_score',
       'review_answer_timestamp', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'order_item_id', 'product_id',
       'seller_id', 'price', 'freight_value', 'seller_zip_code_prefix',
       'seller_state', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'customer_zip_code_prefix', 'customer_state'],
      dtype='object')

In [29]:
#Creating new features with timestamps
df["order_purchase_day"] = df["order_purchase_timestamp"].dt.day
df["order_purchase_month"] = df["order_purchase_timestamp"].dt.month
df["order_approved_day"] = df["order_approved_at"].dt.day
df["order_approved_month"] = df["order_approved_at"].dt.month



In [30]:
df["order_delivered_carrier_day"] = df["order_delivered_carrier_date"].dt.day
df["order_delivered_customer_day"] = df["order_delivered_customer_date"].dt.day

In [31]:
df.drop(["order_purchase_timestamp", "order_approved_at","order_estimated_delivery_date", "review_answer_timestamp" ], axis=1, inplace=True)
df.drop(["order_delivered_carrier_date", "order_delivered_customer_date"], axis=1, inplace= True)

In [32]:
df.head()

order_status  review_score  payment_sequential payment_type  \
0             8             4                 1.0  credit_card   
1             8             4                 3.0      voucher   
2             8             4                 2.0      voucher   
3             8             4                 1.0       boleto   
4             8             5                 1.0  credit_card   

   payment_installments  payment_value  order_item_id  \
0                   1.0          18.12            1.0   
1                   1.0           2.00            1.0   
2                   1.0          18.59            1.0   
3                   1.0         141.46            1.0   
4                   3.0         179.12            1.0   

                         product_id                         seller_id   price  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   29.99   
1  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   29.99   
2  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   29.99   
3  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962  118.70   
4  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2  159.90   

   ...  product_height_cm  product_width_cm customer_zip_code_prefix  \
0  ...                8.0              13.0                     3149   
1  ...                8.0              13.0                     3149   
2  ...                8.0              13.0                     3149   
3  ...               13.0              19.0                    47813   
4  ...               19.0              21.0                    75265   

  customer_state  order_purchase_day  order_purchase_month  \
0             SP                   2                    10   
1             SP                   2                    10   
2             SP                   2                    10   
3             BA                  24                     7   
4             GO                   8                     8   

   order_approved_day  order_approved_month  order_delivered_carrier_day  \
0                 2.0                  10.0                          4.0   
1                 2.0                  10.0                          4.0   
2                 2.0                  10.0                          4.0   
3                26.0                   7.0                         26.0   
4                 8.0                   8.0                          8.0   

   order_delivered_customer_day  
0                          10.0  
1                          10.0  
2                          10.0  
3                           7.0  
4                          17.0  

[5 rows x 29 columns]

In [33]:
df.columns

Index(['order_status', 'review_score', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'order_item_id', 'product_id',
       'seller_id', 'price', 'freight_value', 'seller_zip_code_prefix',
       'seller_state', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'customer_zip_code_prefix', 'customer_state', 'order_purchase_day',
       'order_purchase_month', 'order_approved_day', 'order_approved_month',
       'order_delivered_carrier_day', 'order_delivered_customer_day'],
      dtype='object')

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119151 entries, 0 to 119150
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   order_status                  119151 non-null  int64  
 1   review_score                  119151 non-null  int64  
 2   payment_sequential            119148 non-null  float64
 3   payment_type                  119148 non-null  object 
 4   payment_installments          119148 non-null  float64
 5   payment_value                 119148 non-null  float64
 6   order_item_id                 118318 non-null  float64
 7   product_id                    118318 non-null  object 
 8   seller_id                     118318 non-null  object 
 9   price                         118318 non-null  float64
 10  freight_value                 118318 non-null  float64
 11  seller_zip_code_prefix        118318 non-null  float64
 12  seller_state                  118318 non-nul

In [35]:
df[["seller_state", "customer_state"]]

seller_state customer_state
0                SP             SP
1                SP             SP
2                SP             SP
3                SP             BA
4                SP             GO
...             ...            ...
119146           SP             SP
119147           SP             BA
119148           MG             RJ
119149           MG             RJ
119150           SP             PR

[119151 rows x 2 columns]

In [36]:
seller_state_map = round(df["seller_state"].value_counts()/len(df["seller_state"]),4).to_dict()
customer_state_map = round(df["customer_state"].value_counts()/len(df["customer_state"]),4).to_dict()

In [37]:
# Mapping columns
df["seller_state"] = df["seller_state"].map(seller_state_map)
df["customer_state"] = df["customer_state"].map(customer_state_map)


In [38]:
df.drop(["product_id", "seller_id"], axis=1, inplace=True)

In [39]:
ohe = pd.get_dummies(df["payment_type"])
df = pd.concat([df, ohe], axis=1)

In [40]:
df.head()

order_status  review_score  payment_sequential payment_type  \
0             8             4                 1.0  credit_card   
1             8             4                 3.0      voucher   
2             8             4                 2.0      voucher   
3             8             4                 1.0       boleto   
4             8             5                 1.0  credit_card   

   payment_installments  payment_value  order_item_id   price  freight_value  \
0                   1.0          18.12            1.0   29.99           8.72   
1                   1.0           2.00            1.0   29.99           8.72   
2                   1.0          18.59            1.0   29.99           8.72   
3                   1.0         141.46            1.0  118.70          22.76   
4                   3.0         179.12            1.0  159.90          19.22   

   seller_zip_code_prefix  ...  order_purchase_month order_approved_day  \
0                  9350.0  ...                    10                2.0   
1                  9350.0  ...                    10                2.0   
2                  9350.0  ...                    10                2.0   
3                 31570.0  ...                     7               26.0   
4                 14840.0  ...                     8                8.0   

   order_approved_month  order_delivered_carrier_day  \
0                  10.0                          4.0   
1                  10.0                          4.0   
2                  10.0                          4.0   
3                   7.0                         26.0   
4                   8.0                          8.0   

   order_delivered_customer_day  boleto  credit_card  debit_card  not_defined  \
0                          10.0       0            1           0            0   
1                          10.0       0            0           0            0   
2                          10.0       0            0           0            0   
3                           7.0       1            0           0            0   
4                          17.0       0            1           0            0   

   voucher  
0        0  
1        1  
2        1  
3        0  
4        0  

[5 rows x 32 columns]

In [41]:
df_new = df.dropna(subset=['payment_sequential', 'order_item_id', "order_approved_day", "product_weight_g"])
df_new.head()

order_status  review_score  payment_sequential payment_type  \
0             8             4                 1.0  credit_card   
1             8             4                 3.0      voucher   
2             8             4                 2.0      voucher   
3             8             4                 1.0       boleto   
4             8             5                 1.0  credit_card   

   payment_installments  payment_value  order_item_id   price  freight_value  \
0                   1.0          18.12            1.0   29.99           8.72   
1                   1.0           2.00            1.0   29.99           8.72   
2                   1.0          18.59            1.0   29.99           8.72   
3                   1.0         141.46            1.0  118.70          22.76   
4                   3.0         179.12            1.0  159.90          19.22   

   seller_zip_code_prefix  ...  order_purchase_month order_approved_day  \
0                  9350.0  ...                    10                2.0   
1                  9350.0  ...                    10                2.0   
2                  9350.0  ...                    10                2.0   
3                 31570.0  ...                     7               26.0   
4                 14840.0  ...                     8                8.0   

   order_approved_month  order_delivered_carrier_day  \
0                  10.0                          4.0   
1                  10.0                          4.0   
2                  10.0                          4.0   
3                   7.0                         26.0   
4                   8.0                          8.0   

   order_delivered_customer_day  boleto  credit_card  debit_card  not_defined  \
0                          10.0       0            1           0            0   
1                          10.0       0            0           0            0   
2                          10.0       0            0           0            0   
3                           7.0       1            0           0            0   
4                          17.0       0            1           0            0   

   voucher  
0        0  
1        1  
2        1  
3        0  
4        0  

[5 rows x 32 columns]

In [42]:
df_new.shape

(118280, 32)

In [43]:
import random
import numpy as np
df_new["order_delivered_carrier_day"].fillna(random.choice(df_new[df_new["order_delivered_carrier_day"] != np.nan]["order_delivered_carrier_day"]), inplace =True)
df_new["order_delivered_customer_day"].fillna(random.choice(df_new[df_new["order_delivered_customer_day"] != np.nan]["order_delivered_customer_day"]), inplace =True)


C:\Users\ankit\anaconda3\lib\site-packages\pandas\core\series.py:4523: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [44]:
df_new.isna().sum()

order_status                       0
review_score                       0
payment_sequential                 0
payment_type                       0
payment_installments               0
payment_value                      0
order_item_id                      0
price                              0
freight_value                      0
seller_zip_code_prefix             0
seller_state                       0
product_category_name           1689
product_name_lenght             1689
product_description_lenght      1689
product_photos_qty              1689
product_weight_g                   0
product_length_cm                  0
product_height_cm                  0
product_width_cm                   0
customer_zip_code_prefix           0
customer_state                     0
order_purchase_day                 0
order_purchase_month               0
order_approved_day                 0
order_approved_month               0
order_delivered_carrier_day        0
order_delivered_customer_day       0
b

In [45]:
#Imputing missing values with mean
df_new["product_name_lenght"].fillna(df_new["product_name_lenght"].mean(), inplace=True)
df_new["product_description_lenght"].fillna(df_new["product_description_lenght"].mean(), inplace=True)
df_new["product_photos_qty"].fillna(df_new["product_photos_qty"].mean(), inplace=True)

In [46]:
#drop payment_type
df.drop(["payment_type"], axis=1, inplace=True)

In [47]:
df.product_category_name.nunique()
df.drop(["product_category_name"], axis=1, inplace=True)

In [48]:
df_new.dropna(axis=1, inplace=True)

C:\Users\ankit\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [49]:
X = df_new[['order_status', 'review_score', 'payment_sequential',
       'payment_installments', 'payment_value', 'order_item_id',
       'freight_value', 'seller_zip_code_prefix', 'seller_state',
       'product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm', 'customer_zip_code_prefix',
       'customer_state', 'order_purchase_day', 'order_purchase_month',
       'order_approved_day', 'order_approved_month',
       'order_delivered_carrier_day', 'order_delivered_customer_day', 'boleto',
       'credit_card', 'debit_card', 'not_defined', 'voucher']]
y = df_new.price

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

In [51]:
X.values

array([[8., 4., 1., ..., 0., 0., 0.],
       [8., 4., 3., ..., 0., 0., 1.],
       [8., 4., 2., ..., 0., 0., 1.],
       ...,
       [8., 2., 1., ..., 0., 0., 0.],
       [8., 2., 1., ..., 0., 0., 0.],
       [8., 5., 1., ..., 1., 0., 0.]])

In [ ]:
#y_scaled = scaler.transform(y_train)

In [53]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [52]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

NameError: name 'X_train' is not defined

In [54]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [55]:
model.predict(X_test)

array([34.66872838, 26.56792259, 22.21783621, ..., 63.91391485,
       66.37714042, 54.5746232 ])

In [56]:
model.score(X_test, y_test)

0.6462338481434184

In [57]:
model = LinearRegression()
model.fit(X_train_scaled, y_train)

NameError: name 'X_train_scaled' is not defined

In [58]:
model.score(X_test_scaled, y_test)

NameError: name 'X_test_scaled' is not defined

In [59]:
from sklearn.tree import DecisionTreeRegressor
reg = DecisionTreeRegressor()

In [60]:
reg.fit(X_train, y_train)

DecisionTreeRegressor()

In [61]:
reg.score(X_test, y_test)

0.7374697575508209

In [63]:
from sklearn.ensemble import RandomForestRegressor
reg = RandomForestRegressor()

In [64]:
reg.fit(X_train, y_train)

RandomForestRegressor()

In [65]:
reg.score(X_test, y_test)

0.8812108776547832

We are getting 0.881 r2 score using Random forest algorithm which is a very good score.
The model can be further made more accurate using hyperparameter tuning etc.

In [66]:
from sklearn.svm import SVR
model = SVR()

In [67]:
model.fit(X_train, y_train)

SVR()

In [68]:
model.score(X_test, y_test)

0.034720425537376975

from sklearn.model_selection import RandomizedSearchCV
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
print(random_grid)
rf_random = RandomizedSearchCV(estimator = reg, param_distributions = random_grid, n_iter = 100, cv = 3, verbose=2, random_state=42, n_jobs = -1)
# Fit the random search model
rf_random.fit(X_train, y_train)